### PR TITLE
Port SpatialSoftmax and remove Robomimic dependency

### PR DESCRIPTION
**Pull request by kashyapakshay**
## What this does
This PR introduces a minimal port of the `SpatialSoftmax` op from `robomimic`, removing dependency on the latter. robomimic was also removed from `pyproject.toml`

## How it was tested
Ran diffusion_pusht eval for 500 episodes (w/ batch_size=50 and episode_length=300). metrics:
```
{'avg_sum_reward': 104.51282006196102, 'avg_max_reward': 0.949555460683713, 'pc_success': 64.2, 'eval_s': 1000.3290250301361, 'eval_ep_s': 2.000658050537109}
```
% Success = 64.2%, close to (and slightly better than) what's reported on the hub (63.8%).

Also, had to generate new safetensors for backwards compatibility because we 1) removed temperature params; 2) use one buffer for the normalized grid coordinates instead of the two that robomimic does.

## How to checkout & try? (for the reviewer)
just as before, no extra changes needed.

### CODE REVIEW

**Comment by kashyapakshay** [comment]
~hmm all the backward compatibility tests seems to be passing locally for me, but they fail here for aloha-act and xarm-tdmpc?~

nvm figured it out

Screenshot 2024-05-13 at 6 43 32 PM

**Review by aliberts — COMMENTED** [review]
Awesome, thank you for this!
LGTM but I'll defer to @alexander-soare to approve as he's the expert on this (also I'm not sure we want to keep the temperature params).

File: lerobot/common/policies/diffusion/modeling_diffusion.py
[the remark this answers is not in the document]
Comment (kashyapakshay) [reply lerobot/common/policies/diffusion/modeling_diffusion.py:1]: done!

File: lerobot/common/policies/diffusion/modeling_diffusion.py [hunk 1]
```diff
@@ -289,6 +289,79 @@ def compute_loss(self, batch: dict[str, Tensor]) -> Tensor:
...
+    space using a conv1x1 before computing the softmax.
+    """
+
+    def __init__(self, input_shape, num_kp=None, temperature=1.0, learnable_temperature=False):
+        """
+        Args:
+            input_shape (list): (C, H, W) input feature map shape.
+            num_kp (int): number of keypoints in output. If None, output will have the same number of channels as input.
+            temperature (float): temperature for softmax normalization.
+            learnable_temperature (bool): whether to learn the temperature parameter.
```
[the remark this answers is not in the document]
Comment (kashyapakshay) [reply lerobot/common/policies/diffusion/modeling_diffusion.py]: sounds good, removed those

File: lerobot/common/policies/diffusion/modeling_diffusion.py [hunk 2]
```diff
@@ -289,6 +288,76 @@ def compute_loss(self, batch: dict[str, Tensor]) -> Tensor:
...
+
+    Example: take feature maps of size (512x10x12). We generate a grid of normalized coordinates (10x12x2):
+    -----------------------------------------------------
+    | (-1., -1.)   | (-0.82, -1.)   | ... | (1., -1.)   |
+    | (-1., -0.78) | (-0.82, -0.78) | ... | (1., -0.78) |
+    | ...          | ...            | ... | ...         |
+    | (-1., 1.)    | (-0.82, 1.)    | ... | (1., 1.)    |
+    -----------------------------------------------------
+    We apply channel-wise softmax over the activations (512x120) and compute dot product with the coordinates (120x2)
+    to get expected points of maximal activation (512x2).
```
Comment (alexander-soare) [inline lerobot/common/policies/diffusion/modeling_diffusion.py:307]: nit: to help understand how the next sentence relates to the previous

```suggestion
    This is achieved by applying channel-wise softmax over the activations (512x120) and computing the dot
    product with the coordinates (120x2) to get expected points of maximal activation (512x2).
```

File: lerobot/common/policies/diffusion/modeling_diffusion.py [hunk 3]
```diff
@@ -289,6 +288,76 @@ def compute_loss(self, batch: dict[str, Tensor]) -> Tensor:
...
+
+    Optionally, when num_kp != None, can learn a linear mapping (of shape (num_kp, H, W)) from the input feature maps
+    using a conv1x1 before computing softmax.
```
Comment (alexander-soare) [inline lerobot/common/policies/diffusion/modeling_diffusion.py:310]: nit: connect this to concepts above

```suggestion
    The example above results in 512 keypoints (corresponding to the 512 input channels). We can optionally
    provide num_kp != None to control the number of keypoints. This is achieved by a learnable linear mapping
    (in_channels, H, W) -> (num_kp, H, W).
```

**Review by alexander-soare — APPROVED** [review]
@kashyapakshay this PR is beautiful.
✓ compact, readable code that serves its purpose
✓ excellent documentation that means I can read it and not have to do any detective work or be left with any questions (eg, why you left numpy meshgrid)

**Comment by aliberts** [comment]
Could you upload (here in a comment) your `eval_info.json` from after the change? (if you still have it). This is just to have data to test with on #184

**Comment by kashyapakshay** [comment]
[eval_info.json](https://github.com/huggingface/lerobot/files/15336757/eval_info.json)
yep, here it is @aliberts

**Pull request merged**